tests/fips/openssl/openssl_fips_alglist.pm

```diff
@@ -15,64 +15,76 @@ use base "consoletest";
-          sub { m/^Invalid Hash: 0$/ };
+    my $command = ($type eq 'public_key') ? "$openssl_binary list -public-key-algorithms" : "$openssl_binary list -digest-algorithms";
+    open(my $openssl_output, '-|', $command) or die "Failed to execute $command";
 
-        # List public key algorithms in fips mode
-        # only RSA, DSA, ECDSA, EC DH, CMAC and HMAC are approved in fips mode
-        validate_script_output
-"echo -n 'Invalid Pubkey: '; openssl list-public-key-algorithms | grep '^Name' | sed -e '/RSA/d' -e '/rsa/d' -e '/DSA/d' -e '/dsa/d' -e '/EC/d' -e '/DH/d' -e '/HMAC/d' -e '/CMAC/d' | wc -l",
-          sub { m/^Invalid Pubkey: 0$/ };
-    } elsif (is_sle('<15-SP6')) {
-        eval {
-            validate_script_output
-"echo -n 'Invalid Hash: '; openssl list -digest-algorithms | sed -e '/SHA1/d' -e '/SHA224/d' -e '/SHA256/d' -e '/SHA384/d' -e '/SHA512/d' -e '/DSA/d' -e '/SHA3-224/d' -e '/SHA3-256/d' -e '/SHA3-384/d' -e '/SHA3-512/d' -e '/SHAKE128/d' -e '/SHAKE256/d' | wc -l",
-              sub { m/^Invalid Hash: 0$/ };
-
-            validate_script_output
-"echo -n 'Invalid Pubkey: '; openssl list -public-key-algorithms | grep '^Name' | sed -e '/RSA/d' -e '/rsa/d' -e '/DSA/d' -e '/dsa/d' -e '/EC/d' -e '/DH/d' -e '/HMAC/d' -e '/CMAC/d' | wc -l",
-              sub { m/^Invalid Pubkey: 0$/ };
+    # check for unknown algorithms in the output
+    while (my $line = <$openssl_output>) {
+        chomp $line;
+        if ($type eq 'public_key') {
+            if ($is_openssl3 == 0) {
+                next unless $line =~ /Name:/;
+            } else {
+                next unless $line =~ /@ fips/;
+            }
+        } else {
+            next unless ($is_openssl3 && $line =~ /@ fips/) || ($is_openssl3 == 0);    # only check for 'fips' when using OpenSSL 3.x
         }
-    } else {
-        eval {
-            validate_script_output
-"echo -n 'Invalid Hash: '; openssl list -digest-algorithms | grep fips | sed -e '/SHA1/d' -e '/SHA224/d' -e '/SHA256/d' -e '/SHA384/d' -e '/SHA512/d' -e '/DSA/d' -e '/SHA3-224/d' -e '/SHA3-256/d' -e '/SHA3-384/d' -e '/SHA3-512/d' -e '/SHAKE128/d' -e '/SHAKE256/d' -e '/KECCAK/d' | wc -l",
-              sub { m/^Invalid Hash: 0$/ };
-            validate_script_output
-"echo -n 'Invalid Pubkey: '; openssl list -public-key-algorithms | grep '^Name' | sed -e '/RSA/d' -e '/rsa/d' -e '/DSA/d' -e '/dsa/d' -e '/EC/d' -e '/DH/d' -e '/HMAC/d' -e '/CMAC/d' | wc -l",
-              sub { m/^Invalid Pubkey: 0$/ };
+
+        my $invalid_algo_found = 1;
+        foreach my $valid_algo (@valid_algos) {
+            $invalid_algo_found = 0 if $line =~ /$valid_algo/;
         }
+        die "Error: Invalid algorithm found - $line\n" if ($invalid_algo_found == 1);
+    }
+    close($openssl_output);
+}
+
+sub run_alglist_fips_tests {
+    my $openssl_binary = shift // "openssl";
+    check_algos('public_key', $openssl_binary);
```

Comment:
I'm debated on using this kind of parameters. IMHO two distinct functions like `check_pk_algos()` and `check_hash_algos()` would be clearer and spare some if() statements. 
If you follow this advice, you can also create a separate function to run the command and returns the output, or pass it a block parameter to parse.  
WDYT?